### PR TITLE
Addl fields bugfixes

### PR DESCRIPTION
**Pull request by gcsantos-gpa**

### CODE REVIEW

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/AdditionalFieldForm.tsx [hunk 1]
```diff
@@ -93,19 +93,19 @@ export default function AdditionalFieldForm(props: IProps) {
...
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} />
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} />
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'Searchable'} Label={'Searchable'} Setter={props.Setter} Help={'Make this field searchable?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} Help={'Is this a secure field?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} Help={'Is this an info field?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} Help={'Is this a key field?'} />
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/AdditionalFields/AdditionalFieldForm.tsx:108]:
```suggestion
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'Searchable'} Label={'Searchable'} Setter={props.Setter} Help={'Enable to use this field as a Search Filter.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} Help={'Enable to make this field only visible to admins.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} Help={'Enable to display this field in the Info tab.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} Help={'Enable if this is an External Database Key.'} />
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/CommonComponents/AdditionalFieldsWindow.tsx [hunk 2]
```diff
@@ -301,6 +316,12 @@ function AdditionalFieldsWindow(props: IProps): JSX.Element {
             </div>
             <div className="card-body" style={{ maxHeight: window.innerHeight - 315, overflowY: 'auto' }}>
                 {tableComponent}
+                <ToolTip Show={hoverEdit !== 'None'} Position={'left'} Theme={'dark'} Target={hoverEdit}>
+                    Opens Edit Menu for Key Field
+                </ToolTip>
+                <ToolTip Show={hoverDelete !== 'None'} Position={'left'} Theme={'dark'} Target={hoverDelete}>
+                    Clears Key Field
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/CommonComponents/AdditionalFieldsWindow.tsx:323]:
```suggestion
                    Select Key Field Value
                </ToolTip>
                <ToolTip Show={hoverDelete !== 'None'} Position={'left'} Theme={'dark'} Target={hoverDelete}>
                    Clear Key Field Value
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTable.tsx [hunk 3]
```diff
@@ -88,8 +86,8 @@ export default function ExternalDB(props: { ID: number, Tab: Tab }) {
                 <div className="col">
                     <button className="btn btn-danger pull-right" hidden={record == null}
                         onClick={() => setShowRemove(true)}>Delete Table</button>
-                    <button className="btn btn-light pull-right" hidden={record == null}
-                        onClick={() => { setShowDialog(true); } }>Test Table Query</button>
+                    <button className="btn btn-light pull-right" hidden={record?.ExtDBID == null}
+                        onClick={() => GoToExternalDB()}>Navigate to External Database</button>
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTable.tsx:90]:
```suggestion
                    <button className="btn btn-primary pull-right" hidden={record?.ExtDBID == null}
                        onClick={() => GoToExternalDB()}>Database Details</button>
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx [hunk 4]
```diff
@@ -54,16 +54,16 @@ export default function ExternalDBTableFieldForm(props: IProps) {
...
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} />
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} />
-            <CheckBox<SystemCenter.Types.AdditionalField> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'Searchable'} Label={'Searchable'} Setter={props.Setter} Help={'Make this field searchable?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} Help={'Is this a secure field?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} Help={'Is this an info field?'} />
+            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} Help={'Is this a key field?'} />
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx:66]:
```suggestion
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'Searchable'} Label={'Searchable'} Setter={props.Setter} Help={'Enable to use this field as a Search Filter.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsSecure'} Label={'Secure'} Setter={props.Setter} Help={'Enable to make this field only visible to admins.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsInfo'} Label={'Info'} Setter={props.Setter} Help={'Enable to display this field and its value in the Info tab.'} />
            <CheckBox<SystemCenter.Types.AdditionalFieldView> Record={props.Record} Field={'IsKey'} Label={'Key'} Setter={props.Setter} Help={'Enable if this is an External Database Key.'} />
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx [hunk 5]
```diff
@@ -238,6 +298,7 @@ export default function ExternalDBTableFields(props: { TableName: string, ID: nu
                     { key: 'FieldName', field: 'FieldName', label: 'Name', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
                     { key: 'ParentTable', field: 'ParentTable', label: 'Parent Type', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
                     { key: 'Type', field: 'Type', label: 'Type', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
+                    { key: 'ExternalDB', field: 'ExternalDB', label: 'External DataBase', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx:301]:
```suggestion
                    { key: 'ExternalDB', field: 'ExternalDB', label: 'External Database', headerStyle: { width: 'auto' }, rowStyle: { width: 'auto' } },
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx [hunk 6]
```diff
@@ -195,41 +245,51 @@ export default function ExternalDBTableFields(props: { TableName: string, ID: nu
             </div>
 
             <Modal Title={'Remove External DB Table Field'} Show={showRemove} CallBack={() => setShowRemove(false)} ShowCancel={false} ConfirmText={'Close'} >
-                <button className="btn btn-danger btn-block" onClick={() => { DisassociateField(record); setShowRemove(false); }}>Disassociate Field From This Table</button>
-                <button className="btn btn-danger btn-block" onClick={() => { Delete(); setShowRemove(false); }}>Delete Field Permanently</button>
+                <button className="btn btn-danger btn-block" onClick={() => { setTableStatus('changed'); DisassociateField(record); setShowRemove(false); }}>Disassociate Field From This Table</button>
```
Comment (elwills) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFields.tsx:248]:
```suggestion
                <button className="btn btn-danger btn-block" onClick={() => { setTableStatus('changed'); DisassociateField(record); setShowRemove(false); }}>Remove Field From Table</button>
```

**Review by elwills — APPROVED** [review]

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ByExternalTable.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+                                event.preventDefault()
+                                setRecord({ ...emptyRecord });
+                                setShowNew(true);
+                            }}>Add External Table</button>
+                        </form>
+                    </fieldset>
+                </li>
+            </SearchBar>
+
+            <div style={{ width: '100%', height: 'calc( 100% - 136px)' }}>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ByExternalTable.tsx:98]: this is going to cause problems. We have no guarantee that the searchbar is 136px. wrap the searchbar in a row (if it is not already) and put this in a row and col

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ByExternalTable.tsx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                    sortKey={sortField}
+                    ascending={ascending}
+                    onSort={(d) => {
+                        if (d.colKey === null) return;
+                        dispatch(ExternalDBTablesSlice.Sort({ SortField: d.colField, Ascending: d.ascending }));
+                    }}
+                    onClick={handleSelect}
+                    theadStyle={{ fontSize: 'smaller', display: 'table', tableLayout: 'fixed', width: '100%' }}
+                    tbodyStyle={{ display: 'block', overflowY: 'scroll', maxHeight: window.innerHeight - 300, width: '100%' }}
+                    rowStyle={{ fontSize: 'smaller', display: 'table', tableLayout: 'fixed', width: '100%' }}
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ByExternalTable.tsx:118]: As disucssed `maxHeight` on a `tbody` is bad. put that in a flex row.

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDB.tsx [hunk 9]
```diff
@@ -114,19 +84,15 @@ export default function ExternalDB(props: { ID: number, Tab: Tab }) {
...
-        handle.then(() => {
-            setPopupTitle("Update Results");
-            setRequestStatus('idle');
+            setRequestStatus('error');
         });
         return () => {
             if (handle != null && handle.abort != null) handle.abort();
         };
-    }, [record, setPopupMessage, setRequestStatus, setPopupTitle]);
+    }, [record, setRequestStatus]);
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDB.tsx:95]:
```suggestion
    }, [record]);
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBForm.tsx [hunk 10]
```diff
@@ -1,7 +1,7 @@
 //******************************************************************************************************
 //  ExternalDBForm.tsx - Gbtc
 //
-//  Copyright � 2020, Grid Protection Alliance.  All Rights Reserved.
+//  Copyright � 2020, Grid Protection Alliance.  All Rights Reserved.
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBForm.tsx:4]:
```suggestion
//  Copyright © 2020, Grid Protection Alliance.  All Rights Reserved.
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBForm.tsx [hunk 11]
```diff
@@ -22,45 +22,73 @@
...
+        });
+        handle.fail(() => {
+            setRequestStatus('error');
+        });
+        return () => {
+            if (handle != null && handle.abort != null) handle.abort();
+        };
+    }, [props.Record, setRequestStatus]);
 
+    const ClosePopup = React.useCallback(() => {
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBForm.tsx:71]: that is unneccesarry, state Setters are not changeing Ref so you can use them directly

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBInfo.tsx [hunk 12]
```diff
@@ -64,7 +61,7 @@ export default function ExternalDBInfo(props: { Record: SystemCenter.Types.Exter
                 </div>
             </div>
             <div className="card-body" style={{ height: window.innerHeight - 350, maxHeight: window.innerHeight - 350 }}>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBInfo.tsx:63]: setting `height` to a `px` value is bad layout. Let's fix this all the way up to be a `flex` row

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTable.tsx [hunk 13]
```diff
@@ -88,8 +86,8 @@ export default function ExternalDB(props: { ID: number, Tab: Tab }) {
                 <div className="col">
                     <button className="btn btn-danger pull-right" hidden={record == null}
                         onClick={() => setShowRemove(true)}>Delete Table</button>
-                    <button className="btn btn-light pull-right" hidden={record == null}
-                        onClick={() => { setShowDialog(true); } }>Test Table Query</button>
+                    <button className="btn btn-light pull-right" hidden={record?.ExtDBID == null}
+                        onClick={() => GoToExternalDB()}>Database Details</button>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTable.tsx:90]:
```suggestion
                        onClick={(GoToExternalDB}>Database Details</button>
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx [hunk 14]
```diff
@@ -1,7 +1,7 @@
 //******************************************************************************************************
 //  ExternalDBTableFieldForm.tsx - Gbtc
 //
-//  Copyright � 2023, Grid Protection Alliance.  All Rights Reserved.
+//  Copyright � 2023, Grid Protection Alliance.  All Rights Reserved.
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableFieldForm.tsx:4]:
```suggestion
//  Copyright © 2023, Grid Protection Alliance.  All Rights Reserved.
```

not sure why this keeps happening. someone must have their settings in VS or VSCode set incorrectly

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableInfo.tsx [hunk 15]
```diff
@@ -71,7 +71,7 @@ export default function ExternalDBInfo(props: { Record: SystemCenter.Types.extDB
                 </div>
             </div>
             <div className="card-body" style={{ height: window.innerHeight - 350, maxHeight: window.innerHeight - 350 }}>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/ExternalDB/ExternalDBTableInfo.tsx:73]: no `height` or `maxHeight` please fix to `flex` or `row` where appropriate

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx [hunk 16]
```diff
@@ -204,9 +202,6 @@ function Meter(props: IProps) {
                 <div className={"tab-pane " + (tab == "configurationHistory" ? " active" : "fade")} id="configurationHistory">
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx:202]:
```suggestion
                {tab == "configurationHistory" ?<div className={"tab-pane active" }>
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx [hunk 17]
```diff
@@ -204,9 +202,6 @@ function Meter(props: IProps) {
...
                     <MeterConfigurationHistoryWindow Meter={meter} />
                 </div>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx:204]:
```suggestion
                </div> : null}
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx [hunk 18]
```diff
@@ -204,9 +202,6 @@ function Meter(props: IProps) {
...
-                <div className={"tab-pane " + (tab == "extDB" ? " active" : "fade")} id="extDB">
-                    <ExternalDBUpdate ID={props.MeterID} Type='Meter' Tab={tab} />
-                </div>
                 <div className={"tab-pane " + (tab == "dataRescue" ? " active" : "fade")} id="dataRescue">
                     {dataRescueWindow}
                 </div>
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/Meter/Meter.tsx:207]:
```suggestion
                {tab == "dataRescue" ? <div className={"tab-pane active"}>
                    {dataRescueWindow}
                </div> : null }
```

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/NewMeterWizard/NewMeterWizard.tsx [hunk 19]
```diff
@@ -51,8 +51,12 @@ const trendChannelsStep: number = eventChannelsStep + 1;
 const assetStep: number = trendChannelsStep + 1;
 const connectionStep: number = assetStep + 1;
 const additionalFieldMeterStep: number = connectionStep + 1;
+/* Also undo external turn on in addl
```
Comment (clackner-gpa) [inline Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/NewMeterWizard/NewMeterWizard.tsx:54]: Is the comment still outstanding?

**Review by clackner-gpa — CHANGES_REQUESTED** [review]

File: Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/NewMeterWizard/NewMeterWizard.tsx (hunk 19 above)
Comment (gcsantos-gpa) [reply Source/Applications/SystemCenter/wwwroot/Scripts/TSX/SystemCenter/NewMeterWizard/NewMeterWizard.tsx:54]: Yes, we need to fix elsewhere before we can turn this step back on

**Comment by clackner-gpa** [comment]
Resolving outstanding comments in favor of merging for deployment. @pdinsdale is working on layout fixes in separate Branch.

**Review by clackner-gpa — APPROVED** [review]

**Pull request merged**